drivers/i2c/i2c_rpi_pico_pio.c

```diff
@@ -0,0 +1,567 @@
+	}
+
+	data->config = dev_config;
+	return 0;
+}
+
+static int i2c_pio_get_config(const struct device *dev, uint32_t *config)
+{
+	struct i2c_pio_data *data = dev->data;
+
+	return data->config;
+}
+
+static int i2c_pio_transfer(const struct device *dev, struct i2c_msg *msgs, uint8_t num_msgs,
+			    uint16_t addr)
+{
+	const struct i2c_pio_config *cfg = dev->config;
+	struct i2c_pio_data *data = dev->data;
+
+	PIO pio = pio_rpi_pico_get_pio(cfg->piodev);
+	uint sm = cfg->sm;
+	int rc;
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+
+	for (int i = 0; i < num_msgs; i++) {
+		if (msgs[i].flags & I2C_MSG_READ) {
+			rc = pio_i2c_read_blocking(pio, sm, addr, msgs[i].buf, msgs[i].len,
+						   msgs[i].flags);
+		} else {
+			rc = pio_i2c_write_blocking(pio, sm, addr, msgs[i].buf, msgs[i].len,
+						    msgs[i].flags);
+		}
+		if (rc) {
+			rc = -EIO;
+			goto done;
+		}
+	}
+
+	rc = 0;
+done:
+	k_mutex_unlock(&data->lock);
+	return rc;
+}
+
+static int i2c_pio_recover_bus(const struct device *dev)
+{
+	const struct i2c_pio_config *cfg = dev->config;
+	struct i2c_pio_data *data = dev->data;
+	PIO pio = pio_rpi_pico_get_pio(cfg->piodev);
+	uint sm = cfg->sm;
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+
+	pio_i2c_resume_after_error(pio, sm);
+
+	k_mutex_unlock(&data->lock);
+	return 0;
+}
+
+static const struct i2c_driver_api api = {
+	.configure = i2c_pio_configure,
+	.get_config = i2c_pio_get_config,
+	.transfer = i2c_pio_transfer,
+	.recover_bus = i2c_pio_recover_bus,
+};
+
+#if NUM_PIOS == 2
+static uint8_t i2c_pio_program_offset[NUM_PIOS] = {255, 255};
+static int get_pio_idx(PIO pio)
+{
+	return pio == pio0 ? 0 : 1;
+}
+#elif NUM_PIOS == 3
+static uint8_t i2c_pio_program_offset[NUM_PIOS] = {255, 255, 255};
+static int get_pio_idx(PIO pio)
+{
+	return pio == pio0 ? 0 : pio == pio1 ? 1 : 2;
+}
+#else
+BUILD_ASSERT(0, "Unsupported number of PIOs");
+#endif
+K_MUTEX_DEFINE(i2c_pio_program_loader_mtx);
+
+/**
+ * Load i2c program once per pio, if using multiple instances of this driver.
+ *
+ * @return offset of the program in the PIO.
+ */
+static uint8_t load_program_once(PIO pio)
+{
+	uint8_t offset;
+	int pionum = get_pio_idx(pio);
+
+	k_mutex_lock(&i2c_pio_program_loader_mtx, K_FOREVER);
+	if (i2c_pio_program_offset[pionum] == 255) {
+		i2c_pio_program_offset[pionum] = pio_add_program(pio, &i2c_program);
+	}
+	offset = i2c_pio_program_offset[pionum];
+	k_mutex_unlock(&i2c_pio_program_loader_mtx);
+	return offset;
+}
+
+static int i2c_pio_init(const struct device *dev)
+{
+	const struct i2c_pio_config *cfg = dev->config;
+	struct i2c_pio_data *data = dev->data;
+	const struct pinctrl_state *pinctrl;
+	int rc;
+
+	if (!device_is_ready(cfg->piodev)) {
+		return -ENODEV;
+	}
+
+	__ASSERT(!is_spin_locked(spin_lock_instance(PICO_SPINLOCK_ID_HARDWARE_CLAIM)),
+		 "hardware claim lock should not be locked right now.");
+
+	k_mutex_init(&data->lock);
+	data->config = I2C_SPEED_SET(I2C_SPEED_DT);
+
+	PIO pio = pio_rpi_pico_get_pio(cfg->piodev);
```

Comment:
If you have adopted the old C-style variable declaration at the beginning policy, you should make it consistent throughout.